### This notebook is the analysis for Week 3 assignment: Segmentation and Clustering neighborhoods in Toronto

## Part 1: Getting neighborhood data

Load packages that will be used

In [1]:
import pandas as pd
import numpy as np

Scrape the table from wiki using pandas

In [2]:
df = pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")[0]

The dataframe already consists of three columns: PostalCode, Borough, and Neighborhood

In [3]:
df.head(10)

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront
5      M6A        North York  Lawrence Heights
6      M6A        North York    Lawrence Manor
7      M7A  Downtown Toronto      Queen's Park
8      M8A      Not assigned      Not assigned
9      M9A      Queen's Park      Not assigned

Only process the cells that have an assigned borough. The following line removes rows with borough that is Not assigned.

In [4]:
df = df[df["Borough"] != 'Not assigned']

In [5]:
df.head(10)

Postcode           Borough     Neighbourhood
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M6A        North York  Lawrence Heights
6       M6A        North York    Lawrence Manor
7       M7A  Downtown Toronto      Queen's Park
9       M9A      Queen's Park      Not assigned
10      M1B       Scarborough             Rouge
11      M1B       Scarborough           Malvern
13      M3B        North York   Don Mills North

The next 2 requirements are:
- More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.

- If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.

I find doing the second one more reasonable before merging neighborhoods with the same postal code and borough.

Assign the same Borough to Not assigned neighborhood

In [6]:
df['Neighbourhood'] = np.where(df['Neighbourhood'] == 'Not assigned', df['Borough'], df['Neighbourhood'])

Merge neughborhoods with the same postal code and borough

In [7]:
df = df.groupby(['Postcode','Borough'], as_index = False).agg({'Neighbourhood': ', '.join})

In [8]:
df.head(10)

Postcode      Borough                                    Neighbourhood
0      M1B  Scarborough                                   Rouge, Malvern
1      M1C  Scarborough           Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough                Guildwood, Morningside, West Hill
3      M1G  Scarborough                                           Woburn
4      M1H  Scarborough                                        Cedarbrae
5      M1J  Scarborough                              Scarborough Village
6      M1K  Scarborough      East Birchmount Park, Ionview, Kennedy Park
7      M1L  Scarborough                  Clairlea, Golden Mile, Oakridge
8      M1M  Scarborough  Cliffcrest, Cliffside, Scarborough Village West
9      M1N  Scarborough                      Birch Cliff, Cliffside West

Final cleaned dataframe size:

In [9]:
df.shape

(103, 3)

Number of row: 103

For part 3, github doesn't display the folio map correctly, please see it in the page below: https://nbviewer.jupyter.org/github/nhohung/github_example/blob/master/SegmentCluster_Toronto_Week3.ipynb

## Part 2: Getting neighborhood coordinates

Load geographical coordinates

In [10]:
data = pd.read_csv('http://cocl.us/Geospatial_data')

In [11]:
data.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Rename Postal Code to match neighborhood dataframe column Postcode before merging

In [12]:
data.rename(columns={"Postal Code": "Postcode"}, inplace = True)

Merge with neighborhood data

In [13]:
df = pd.merge(df, data, on="Postcode")

In [14]:
df.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

For part 3, github doesn't display the folio map correctly, please see it in the page below: https://nbviewer.jupyter.org/github/nhohung/github_example/blob/master/SegmentCluster_Toronto_Week3.ipynb

## Part 3: Cluster neighborhood

In [15]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [16]:
CLIENT_ID = 'HRQEKGN0SL2TOMK2PKTGEP4ONM3ZM3ZWLPKOERVPYHNISX5M' # your Foursquare ID
CLIENT_SECRET = 'YJZ4KJ5H3CC4V1DW3Z2Y0L5UTQPKSTJRU1HSSVOE2VKLU3OY' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

In [17]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
#         print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Postcode', 
                  'Latitude', 
                  'Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now, let's get the top 100 venues for each neighborhoods within a radius of 500 meters.
#### NOTE: this run for each Postcode, not each neighborhood

In [18]:
toronto_venues = getNearbyVenues(names=df['Postcode'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )

Analyze each postcode

In [19]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Postcode'] = toronto_venues['Postcode'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

# toronto_onehot.head()

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [20]:
toronto_grouped = toronto_onehot.groupby('Postcode').mean().reset_index()
# toronto_grouped

In [21]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [22]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Postcode']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Postcode'] = toronto_grouped['Postcode']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Postcode 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0      M1B  Fast Food Restaurant      Department Store           Event Space   
1      M1C                   Bar           Yoga Studio            Donut Shop   
2      M1E     Electronics Store   Rental Car Location                   Spa   
3      M1G           Coffee Shop     Korean Restaurant           Yoga Studio   
4      M1H  Caribbean Restaurant                  Bank   Fried Chicken Joint   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0  Ethiopian Restaurant   Empanada Restaurant     Electronics Store   
1    Dim Sum Restaurant                 Diner        Discount Store   
2        Breakfast Spot           Pizza Place    Mexican Restaurant   
3             Drugstore                 Diner        Discount Store   
4       Thai Restaurant    Athletics & Sports           Gas Station   

         7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
1                      Dog Run      Doner Restaurant             Drugstore   
2                 Intersection        Medical Center               Dog Run   
3                      Dog Run      Doner Restaurant            Donut Shop   
4             Hakka Restaurant                Lounge                Bakery   

        10th Most Common Venue  
0                   Donut Shop  
1               Farmers Market  
2           Dim Sum Restaurant  
3  Eastern European Restaurant  
4  Eastern European Restaurant

Now run k-means

In [23]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Postcode', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
# kmeans.labels_

Now merge the class to our dataframe

In [24]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Postcode'), on='Postcode')

toronto_merged.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353             2.0  Fast Food Restaurant      Department Store   
1 -79.160497             1.0                   Bar           Yoga Studio   
2 -79.188711             2.0     Electronics Store   Rental Car Location   
3 -79.216917             2.0           Coffee Shop     Korean Restaurant   
4 -79.239476             2.0  Caribbean Restaurant                  Bank   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Event Space  Ethiopian Restaurant   Empanada Restaurant   
1            Donut Shop    Dim Sum Restaurant                 Diner   
2                   Spa        Breakfast Spot           Pizza Place   
3           Yoga Studio             Drugstore                 Diner   
4   Fried Chicken Joint       Thai Restaurant    Athletics & Sports   

  6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
0     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   
1        Discount Store                      Dog Run      Doner Restaurant   
2    Mexican Restaurant                 Intersection        Medical Center   
3        Discount Store                      Dog Run      Doner Restaurant   
4           Gas Station             Hakka Restaurant                Lounge   

  9th Most Common Venue       10th Most Common Venue  
0             Drugstore                   Donut Shop  
1             Drugstore               Farmers Market  
2               Dog Run           Dim Sum Restaurant  
3            Donut Shop  Eastern European Restaurant  
4                Bakery  Eastern European Restaurant

Remove Nan values in Cluster Labels and convert it from Float to Int

In [25]:
toronto_merged = toronto_merged.dropna()
toronto_merged["Cluster Labels"] = toronto_merged["Cluster Labels"].astype(int)

For the map below, github doesn't display the folio map correctly, please see it in the following page: https://nbviewer.jupyter.org/github/nhohung/github_example/blob/master/SegmentCluster_Toronto_Week3.ipynb

In [26]:
# create map
map_clusters = folium.Map(location=[43.6532, -79.3832], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Postcode'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + '. Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
#         color=cluster-1,
        fill=True,
        fill_color=rainbow[cluster-1],
#         fill_color=cluster,
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Again, please see the map in the following page: https://nbviewer.jupyter.org/github/nhohung/github_example/blob/master/SegmentCluster_Toronto_Week3.ipynb